torch/optim/adagrad.py

```diff
@@ -64,6 +61,13 @@ def step(self, closure=None):
                 grad = p.grad.data
                 state = self.state[p]
 
+                # state initialization
+                if len(state) == 0:
+                    state['step'] = 0
+                    state['sum'] = torch.full_like(p.data, group['initial_accumulator_value'])
+                    if self._is_share_memory:
+                        state['sum'].share_memory_()
```

Comment:
Because this pr focuses on the lazy init, so I would consider the following situation where the flag is needed.

```python
opt = Adagrad(params)
opt.share_memory()
opt.add_parameters()

# do something

```

I think the added parameters should behave the same as original parameters, but ignoring the share_memory request for newly added parameters should be fine.
@apaszke What's your opinion?

Reply:
@apaszke The `set_state` and  `(load_)state_dict` methods are updated, can U review this again?